openvpn-wire-protocol.xml

```diff
@@ -1698,9 +1907,83 @@ authenticated_data_v2 = opcode| key_id | peer_id | packet_id | payload
         </t>
 
         <t>
-          The tag size is always 128 bit (16 bytes). (Same size as in TLS). If the
-          protocol flag for AEAD tag at the end is in effect, the AEAD tag is at the end
-          of the packet rather than at the start.
+          The tag size is always 128 bit (16 bytes). (Same size as in TLS).
+        </t>
+      </section>
+    <section anchor="data_epoch" title="AEAD encrypted data channel packet with extended packet id and epoch keys">
+      <t>
+        This is the modern version of the AEAD data format in OpenVPN. The differences to the older
+        format are:
+        <ul>
+          <li>AEAD tag at the end of the packet</li>
+          <li>64 bit packet id </li>
+          <li>IV constructed with XOR instead of concatenation </li>
+          <li>Support of multiple key generations (epochs) </li>
+          <li>Only defined for DATA_V2 packets</li>
+        </ul>
+
+        The packet format with 64bit counter looks like this:
+          <figure>
+            <sourcecode>
+struct aead_packet {
+    int opcode:5;
+    int key_id:3;
+    int peer_id:24;
+    uint_64_t packet_id;
+    uint8_t* encrypted_payload;
+    uint8_t[16] authentication_tag;    
+}
+            </sourcecode>
+          </figure>
+        </t>
+
+        <t>
+          The packet id is split into 16 bit bits of epoch and 48-bit of counter for
+          that epoch. Each epoch uses a different a encryption/decryption key and a
+          different implicit_iv. The epoch selects the key to use for decryption.
+          <figure>
+            <sourcecode>
+struct packet_id {
+    uint epoch:16;
+    uint epoch_counter:48;
+}
+            </sourcecode>
+          </figure>
+        </t>
+
+        <t>
+          The epoch_counter of the packet_id together with the implicit IV for the epoch
+          forms the IV for decryption. The epoch_counter is also used for the replay
+          protection.  Implementations are encouraged to a have a per epoch replay protection.
+          Using the whole packet_id as a flat 64bit counter for replay protection will work as
+          well but will reject reordered packets from the older epochs. A (valid) packet
+          of the new the epoch will be seen as a large gap in the counter that and that will disallow any
+          allow any packet from an earlier epoch.
+        </t>
+
+        <t>
+          The 96bit IV is constructed by using the 96 bit of implicit IV belonging to the epoch
+          and doing an XOR with the 64 bit packet id of both epoch and epoch_counter.
+          <sourcecode>IV = (epoch | epoch_counter) XOR implicit_iv_epoch</sourcecode>
```

Comment:
Yes. It might not be needed but it is one of thing that allows us to not have the real IV exposed. And of the main things why we do this, is that TLS 1.3 does the same thing: https://datatracker.ietf.org/doc/html/rfc8446#section-5.3

Reply:
Makes sense. I can't see any button to mark it resolved, but I consider this comment resolved.